File: src/c_pubsub_jo/src/subscriber_member_function.cpp
#include <functional>
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "interface_jojo/msg/num.hpp"

using std::placeholders::_1;

class MinimalSubscriber : public rclcpp::Node
{
  public:
    MinimalSubscriber()
    : Node("minimal_subscriber")
    {
      subscription_ = this->create_subscription<interface_jojo::msg::Num>(
      "topic", 10, std::bind(&MinimalSubscriber::topic_callback, this, _1));
    }

  private:
    void topic_callback(const interface_jojo::msg::Num & msg) const
    {
      RCLCPP_INFO_STREAM(this->get_logger(), "I heard: '" << msg.num << " ' ");
    }
    rclcpp::Subscription<interface_jojo::msg::Num>::SharedPtr subscription_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<MinimalSubscriber>());
  rclcpp::shutdown();
  return 0;
}

File: src/p_pubsub/p_pubsub/subscriber_member_function.py
import rclpy
from rclpy.node import Node

from interface_jojo.msg import Num


class MinimalSubscriber(Node):

    def __init__(self):
        super().__init__('minimal_subscriber')
        self.subscription = self.create_subscription(
            Num,
            'topic',
            self.listener_callback,
            10)
        self.subscription  # prevent unused variable warning

    def listener_callback(self, msg):
        self.get_logger().info('heh? i heard: "%d"' % msg.num)


def main(args=None):
    rclpy.init(args=args)

    minimal_subscriber = MinimalSubscriber()

    rclpy.spin(minimal_subscriber)

    # Destroy the node explicitly
    # (optional - otherwise it will be done automatically
    # when the garbage collector destroys the node object)
    minimal_subscriber.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
